File: src-server/src/db/schemas/toolset.py
from dais_sdk import LocalServerParams, RemoteServerParams
from . import DTOBase
from ..models.toolset import ToolsetType

class ToolBase(DTOBase):
    name: str
    description: str
    is_enabled: bool
    auto_approve: bool

class ToolRead(ToolBase):
    id: int
    toolset_id: int
    internal_key: str

class ToolUpdate(DTOBase):
    id: int
    name: str | None
    is_enabled: bool | None
    auto_approve: bool | None

class ToolsetBase(DTOBase):
    name: str
    type: ToolsetType
    params: LocalServerParams | RemoteServerParams | None
    is_enabled: bool

class ToolsetRead(ToolsetBase):
    id: int
    internal_key: str
    tools: list[ToolRead]

class ToolsetCreate(DTOBase):
    name: str
    type: ToolsetType
    params: LocalServerParams | RemoteServerParams

class ToolsetUpdate(DTOBase):
    name: str | None
    params: LocalServerParams | RemoteServerParams | None
    is_enabled: bool | None
    tools: list[ToolUpdate] | None

File: src-frontend/src/types/toolset.ts
export type McpToolsetStatus =
  | "connecting"
  | "connected"
  | "disconnected"
  | "error";

export type LocalServerParams = {
  command: string;
  args?: string[];
  env?: Record<string, string>;
  cwd?: string;
  encoding?: string;
  encoding_error_handler?: "strict" | "ignore" | "replace";
};

export type OAuthParams = {
  oauth_scopes?: string[];
  oauth_timeout?: number;
};

export type RemoteServerParams = {
  url: string;
  bearer_token?: string;
  oauth_params?: OAuthParams;
  http_headers?: Record<string, string>;
};

export type ToolsetParams = LocalServerParams | RemoteServerParams;

export type ToolsetType = ToolsetBase["type"];

export type ToolsetBrief = {
  id: number;
  name: string;
  type: ToolsetType;
  // only available for MCP toolsets
  status: McpToolsetStatus | null;
};

export type ToolRead = {
  id: number;
  toolset_id: number;
  name: string;
  description: string;
  internal_key: string;
  is_enabled: boolean;
  auto_approve: boolean;
};

export type ToolUpdate = {
  id: number;
  name?: string;
  is_enabled?: boolean;
  auto_approve?: boolean;
};

export type ToolsetBase = {
  name: string;
} & (
  | { type: "built_in"; params: null }
  | { type: "mcp_local"; params: LocalServerParams }
  | { type: "mcp_remote"; params: RemoteServerParams }
);

export type ToolsetRead = ToolsetBase & {
  id: number;
  internal_key: string;
  is_enabled: boolean;
  tools: ToolRead[];
};

export type ToolsetCreate = ToolsetBase;

export type ToolsetUpdate = Partial<
  ToolsetBase & {
    is_enabled: boolean;
    tools: ToolUpdate[];
  }
>;
